lib/pbench/cli/agent/commands/tools/list.py

```diff
@@ -0,0 +1,142 @@
+"""
+pbench-list-tools
+
+This script can list all tools from all groups, list tools from a specific
+group, or list which gourps contain a specific tool
+
+"""
+
+import logging
+import sys
+
+import click
+
+from pbench.agent.utils import setup_logging
+from pbench.cli.agent import CliContext, pass_cli_context
+from pbench.cli.agent.commands.tools.base import ToolCommand
+from pbench.cli.agent.options import common_options
+
+LOG = logging.getLogger(__name__)
+
+
+class ListTools(ToolCommand):
+    """ List registered Tools """
+
+    def __init__(self, context):
+        super(ListTools, self).__init__(context)
+
+        setup_logging(debug=False, logfile=self.pbench_log)
+
+    def execute(self):
+        if not self.pbench_run.exists():
+            LOG.warn("The %s directory does not exist", self.pbench_run)
+            return 0
+
+        # list tools in one of all groups
+        if self.context.group:
+            groups = self.context.group
+        else:
+            groups = self.groups
+
+        if not self.context.name:
+            host_tools = {}
+            for group in groups:
+                host_tools[group] = {}
+                for path in self.gen_tools_group_dir(group).glob("*/**"):
+                    if self.context.with_option:
+                        host_tools[group][path.name] = [
+                            p.read_text() for p in self.tools(path)
+                        ]
+                    else:
+                        host_tools[group][path.name] = [p for p in self.tools(path)]
+            if host_tools:
+                for k, v in host_tools.items():
+                    for h, t in v.items():
+                        print("%s: %s %s" % (k, h, t))
+        else:
+            # List the groups which include this tool
+            group_list = []
+            for group in groups:
+                tg_dir = self.gen_tools_group_dir(group)
+                if not tg_dir.exists():
+                    LOG.error("bad or missing tool gorup %s", group)
```

Comment:
Not super-important since it's a log message rather than user output, but "gorup" could be corrected at some point!